### DISCUSSION TITLE
Code extend for use with local storage.

### DISCUSSION THREAD

**Discussion by pdchristian** [nikolamilosevic86/local-genAI-search#7] [General]
Hello Nikola,

thank you for creating this code sample.

The last days I have been using it to da a data ingest with my files saved on a LAN server (Synology).

After having done some minor improvements, I could ingest a folder with approximately 15 GB of mixed file type data (of course only txt, pdf and docx were ingested).

What did I do to make it work?
- Added recursive document pull
- Ingest fix: skipped all files with an "~" in the file name
- Ingest fix: skipped all txt files with character encoding problems
- Ingest fix: skipped all pdf files with encryption problems
- Updated prepromt to answer in German
- Added PyCryptodome to the requirements.txt
- Changed model_kwargs = {'device': 'cpu'} to CUDA --> did this change anything?
- Streamlit: Create answer already with return instead of button

Are you interested in the code? Can I contribute here, or should I upload it as a separate repository?

Best regards

Chrstian

**Comment by nikolamilosevic86** [comment]
Hello @pdchristian ,

Thank you for working on the code and getting it further. Yes, I would be interested in the code and what you did, many changes make sense to me, although I would not probably implement answering in German in the main branch, but everything else is absolutely something that makes sense adding in (probably cuda or apple silicon detection should be added and dynamically change it, depending on what people have installed). 

It would be great if you could commit the code as a new branch and let me know. And then I will have a look and merge what makes sense

Best wishes,
Nikola

**Comment by pdchristian** [comment]
Hello Nikola,
yes, I can upload my code as a new branch. But I think you have to give the rights to create a new branch.

Regarding the answer language in the prepromt: I would propose to add a language setting to the streamlit UI. 

Best regards,

Christian

**Comment by nikolamilosevic86** [comment]
Hello @pdchristian,

I have added elements of your code and added a dynamic selection of CUDA, Apple Silicon, or CPU depending on what people have. I will still work on language selection a bit later. If you would like your name to be stated as contributor, please let me know your full name and I can link it to your profile

**Comment by pdchristian** [comment]
Hello Nikola,

yes, it would be nice to have me listed there.

My name is: Christian Barth-Popp


Best regards

Christian

Von: Nikola Milosevic ***@***.***>
Gesendet: Montag, 19. August 2024 12:14
An: nikolamilosevic86/local-genAI-search ***@***.***>
Cc: pdchristian ***@***.***>; Mention ***@***.***>
Betreff: Re: [nikolamilosevic86/local-genAI-search] Code extend for use with local storage. (Discussion #7)


Hello @pdchristian<https://github.com/pdchristian>,

I have added elements of your code and added a dynamic selection of CUDA, Apple Silicon, or CPU depending on what people have. I will still work on language selection a bit later. If you would like your name to be stated as contributor, please let me know your full name and I can link it to your profile

-

**Comment by pdchristian** [comment]
I have actually also added the 
Try:
and
except Excecption as exc: 
to the pptx area.

**Comment by pdchristian** [comment]
Hello Nikola,

I am currently trying to add the language setting in Streamlit to the code.
The problem is, that I do not manage to update the language variable (for the prepromt) at the "async def ask_localai(Item:Item):"

How could we send the language variable to the function to change the language via streamlit dropdown?

Would it be possible to add the value to the json.dumps?

if st.button("Ask a question") or not question=="":
    st.write("The current question is \"", question+"\"")
    url = "http://127.0.0.1:8000/ask_localai"

    payload = json.dumps({
      "query": question
    })
    headers = {
      'Accept': 'application/json',
      'Content-Type': 'application/json'
    }

I added a dropdown for the language to streamlit:

st.session_state.language_ai = st.sidebar.selectbox(
    'Language',
    ('English', 'German', 'French'),
    index = 0,
    key="st.session_state.language_ai",
    on_change=None
)


Thanks in advance

Christian

**Comment by pdchristian** [comment]
And I have done two further improvements:

1) New version of HuggingFaceEmbeddings
replaced: 
from langchain_community.embeddings import HuggingFaceEmbeddings
with:
from langchain_huggingface import HuggingFaceEmbeddings

2) skip Ingest problems for PPT:

        elif file.endswith(".pptx"):
            print("indexing " + file)
            try: # added by pdchristian to catch decryption error
                file_content = getTextFromPPTX(file)
            except Exception as exc:# added by pdchristian to catch decryption error
                file_content = "Empty due to extraction error."  # added by pdchristian to catch decryption error
                
Shall I create a new branch with those changes, so you can easily integrate it into the main branch?
